clang/lib/CodeGen/CGLoopInfo.cpp

```diff
@@ -635,6 +635,17 @@ void LoopInfoStack::push(BasicBlock *Header, clang::ASTContext &Ctx,
         Option = LoopHintAttr::UnrollCount;
         State = LoopHintAttr::Numeric;
       }
+    } else if (HLSLLoopHint) {
+      ValueInt = HLSLLoopHint->getDirective();
+      if (HLSLLoopHint->getSemanticSpelling() ==
+          HLSLLoopHintAttr::Spelling::Microsoft_unroll) {
+        if (ValueInt == 0)
+          State = LoopHintAttr::Enable;
+        if (ValueInt > 0) {
+          Option = LoopHintAttr::UnrollCount;
```

Comment:
I think the `[unroll]` directive in hlsl does not map exactly to the `llvm.loop.unroll.count` metadata interpretation. One difference is that the `[unroll(N)]` will force the iteration count even if the loop bounds is actually larger. So if the loop has an upper bound of `10` and we write `[unroll(3)]` it will force the loop to only execute 3 times.

https://godbolt.org/z/bT3EdcMKs

I don't think the llvm unroll count metadata works the same way. Maybe we want to not support that behavior, but it seems like it is something we should document and maybe apply to HLSL202x if we intend to change the behavior.

Reply:
You should think of the HLSL argument as manually specifying the maximum iterations for the loop.

Reply:
@dmpots you are correct. It's very likely we will pursue a breaking change based on the feedback I've been gathering throughout the day. Since HLSL202x  applies for both clang-dxc and an dxc It may be worth making a similar change in DXC to allow us to measure how much of a breaking change this would be via experiments before we come to a final decision.

Reply:
That sounds like a good path forward to me. Change looks good overall.